src/gui/toolbarMenubar/ToolWindowOpacitySlider.cpp

```diff
@@ -0,0 +1,130 @@
+        pImpl{std::make_unique<Impl>(false, std::move(iconNameHelper), control)} {}
+
+auto ToolWindowOpacitySlider::getToolDisplayName() -> std::string { return _("Window Opacity Slider"); }
+
+auto ToolWindowOpacitySlider::createItem(bool horizontal) -> GtkToolItem* { return pImpl->createItem(horizontal); }
+
+auto ToolWindowOpacitySlider::createTmpItem(bool horizontal) -> GtkToolItem* {
+    return pImpl->createTmpItem(horizontal);
+}
+
+void ToolWindowOpacitySlider::enable(bool enabled) { return pImpl->setEnabled(enabled); }
+
+auto ToolWindowOpacitySlider::newItem() -> GtkToolItem* { return pImpl->newItem(); }
+
+auto ToolWindowOpacitySlider::getNewToolIcon() -> GtkWidget* { return pImpl->getNewToolIcon(); }
+
+
+ToolWindowOpacitySlider::Impl::Impl(bool horizontal, IconNameHelper iconNameHelper, Control* control):
+        horizontal_{horizontal}, iconNameHelper_{iconNameHelper}, control_{control} {}
+
+auto ToolWindowOpacitySlider::Impl::createItem(bool horizontal) -> GtkToolItem* {
+    horizontal_ = horizontal;
+    toolItem_ = newItem();
+    g_object_ref(toolItem_);
+
+    return toolItem_;
+}
+
+auto ToolWindowOpacitySlider::Impl::createTmpItem(bool horizontal) -> GtkToolItem* {
+    GtkToolItem* item = newItem();
+    g_object_ref(item);
+
+    gtk_widget_show_all(GTK_WIDGET(item));
+    return item;
+}
+
+auto ToolWindowOpacitySlider::Impl::newItem() -> GtkToolItem* {
+    GtkToolItem* item = gtk_tool_item_new();
+
+    if (slider_) {
+        // TODO: Who owns slider_? Do we need to do a g_object_unref here?
+        g_signal_handlers_disconnect_by_func(slider_, (void*)(sliderChanged), this);
+    }
+
+    // TODO: Scale fn.
+    double sliderMin = MIN_OPACITY;
+    double sliderMax = MAX_OPACITY;
+    double sliderStep = OPACITY_STEP;
+
+    auto orientation = horizontal_ ? GTK_ORIENTATION_HORIZONTAL : GTK_ORIENTATION_VERTICAL;
+    slider_ = gtk_scale_new_with_range(orientation, sliderMin, sliderMax, sliderStep);
+
+    g_signal_connect(slider_, "value-changed", G_CALLBACK(sliderChanged), this);
+
+    // TODO: Fix duplicated code in ToolZoomSlider
+    if (horizontal_) {
+        gtk_widget_set_size_request(GTK_WIDGET(slider_), 120, 16);
+    } else {
+        gtk_widget_set_size_request(GTK_WIDGET(slider_), 16, 120);
+    }
+
+    MainWindow* window = control_->getWindow();
+    if (window != nullptr) {
+        gtk_range_set_value(GTK_RANGE(slider_), window->getOpacity());
+    } else {
+        gtk_range_set_value(GTK_RANGE(slider_), 1.0);
+    }
+
+    gtk_container_add(GTK_CONTAINER(item), slider_);
+
+    return item;
+}
+
+auto ToolWindowOpacitySlider::Impl::getNewToolIcon() const -> GtkWidget* {
+    return gtk_image_new_from_icon_name(iconNameHelper_.iconName("window-opacity-slider").c_str(),
+                                        GTK_ICON_SIZE_SMALL_TOOLBAR);
+}
+
+void ToolWindowOpacitySlider::Impl::setEnabled(bool enabled) {
+    if (slider_) {
+        gtk_widget_set_sensitive(GTK_WIDGET(slider_), enabled);
```

Comment:
What should we do if we can't set the window's opacity ([`gtk_widget_is_composited`](https://developer.gnome.org/gtk3/stable/GtkWidget.html#gtk-widget-is-composited) returns `false` for the main window)?

One option could be to `gtk_widget_set_sensitive(GTK_WIDGET(slider_), false)` and display a warning when the user hovers over the slider.

Reply:
That sounds like a reasonable solution.